src/app/screens/swap/useAlexSponsoredTransaction.ts

```diff
@@ -24,8 +34,35 @@ export const useAlexSponsoredTransaction = (userOverrideSponsorValue: boolean) =
   const sponsorTransaction = async (signed: StacksTransaction) =>
     alexSDK.broadcastSponsoredTx(signed.serialize().toString('hex'));
 
-  const { data: stxPendingTxData } = useStxPendingTxData();
-  const hasPendingTransactions = stxPendingTxData?.pendingTransactions?.length > 0;
+  const getAccountCurrentNonce = async () => {
+    let nonce = 0;
+    const confirmedTransactions: StxTransactionListData = await getConfirmedTransactions({
+      stxAddress,
+      network: selectedNetwork,
+    });
+
+    if (confirmedTransactions && confirmedTransactions.transactionsList.length > 0) {
+      nonce = confirmedTransactions.transactionsList[0].nonce + 1;
+    }
+    const currentPendingNonce =
+      (stxPendingTxData?.pendingTransactions ?? []).reduce(
+        (maxNonce, transaction) => Math.max(maxNonce, transaction?.nonce ?? 0),
+        0,
+      ) + 1;
+
+    return Math.max(currentPendingNonce, nonce);
+  };
```

Comment:
this is logic that should be going into xverse-core and unit tested, because it is generic enough to be reusable and not limited to swaps.

make it a pure function (no side effects, no fetching), so it's easier to unit test.

see this PR for example of a unit tested pure function: https://github.com/secretkeylabs/xverse-core/pull/205/files

Reply:
thanks for suggesting the core function🙏 . I refactored the logic to filter out dead pending transactions to a simpler one. Please take a look